### ISSUE TITLE
.Net: HandlebarsPlanner "hallucinates" helpers that actually *do* exist (and sometimes work) when using LLamaSharp

### ISSUE THREAD

**Issue by godefroi** [microsoft/semantic-kernel#4751]
**Describe the bug**
I am attempting to use plugin function with a local language model using the LLamaSharp implementation of `IChatCompletionService` and am having trouble. I have had two different results across two runs; the first time I got an exception:
```
Unhandled exception. Microsoft.SemanticKernel.KernelException: [HallucinatedHelpers] The plan references hallucinated helpers: Helper 'unless'
```

The generated Handlebars template looked like this:
```
{{!-- Step 1: Initialize variables --}}
{{set "currentState" ""}}
{{set "newState" false}}

{{!-- Step 2: Get current state of the light --}}
{{#if (LightPlugin-GetState)}}
  {{set "currentState" (LightPlugin-GetState)}}
{{/if}}

{{!-- Step 3: Change the state of the light if it's currently off --}}
{{#if (unless currentState)}}
  {{set "newState" true}}
  {{set "result" (LightPlugin-ChangeState newState)}}
  {{json result}}
{{/if}}
```

The second time I ran the code, I got a different template generated:
```
{{!-- Step 1: Initialize variables --}}
{{set "currentState" ""}}
{{set "newState" false}}

{{!-- Step 2: Get current state of the light --}}
{{#if (LightPlugin-GetState)}}
  {{set "currentState" (LightPlugin-GetState)}}
{{/if}}

{{!-- Step 3: Change state of the light if it's currently off --}}
{{#unless (equals currentState true)}}
  {{set "newState" true}}
  {{set "result" (LightPlugin-ChangeState newState)}}
{{else}}
  {{json "The lights are already on."}}
{{/unless}}

{{!-- Step 4: Output the result --}}
{{json (concat "The lights have been " (if (equals currentState false) "turned on" "remain unchanged"))}}
```

Note that this plan, while different, still references the `unless` handler. Nevertheless, this plan was executed successfully (as shown in the full output below). After the function was executed, an exception was thrown indicating that `if` is a hallucinated handler:
```
Microsoft.SemanticKernel.KernelException: [HallucinatedHelpers] The plan references hallucinated helpers: Helper 'if'
```

**Platform**
 - OS: Windows
 - IDE: VS Code
 - Language: C#
 - Source: NuGet packages:
   - Microsoft.SemanticKernel version 1.0.1
   - Microsoft.SemanticKernel.Planners.Handlebars version 1.0.1-preview
   - Microsoft.SemanticKernel.Planners.Core version 1.0.0-beta8
   - LlamaSharp version 0.9.1
   - LlamaSharp.Backend.Cuda12 version 0.9.1
   - LLamaSharp.semantic-kernel version 0.9.1

**Additional context**
Here is the output of the program. The line "[Light is now on]" is coming from my plugin method, and indicates that the function is being called correctly:

```
info: Microsoft.SemanticKernel.Planning.Handlebars.HandlebarsPlanner[0]
      Plan creation duration: 11.0676889s.
{{!-- Step 1: Initialize variables --}}
{{set "currentState" ""}}
{{set "newState" false}}

{{!-- Step 2: Get current state of the light --}}
{{#if (LightPlugin-GetState)}}
  {{set "currentState" (LightPlugin-GetState)}}
{{/if}}

{{!-- Step 3: Change state of the light if it's currently off --}}
{{#unless (equals currentState true)}}
  {{set "newState" true}}
  {{set "result" (LightPlugin-ChangeState newState)}}
{{else}}
  {{json "The lights are already on."}}
{{/unless}}

{{!-- Step 4: Output the result --}}
{{json (concat "The lights have been " (if (equals currentState false) "turned on" "remain unchanged"))}}
info: Microsoft.SemanticKernel.Planning.Handlebars.HandlebarsPlan[0]
      Plan execution started.
info: GetState[0]
      Function GetState invoking.
info: GetState[0]
      Function GetState succeeded.
info: GetState[0]
      Function completed. Duration: 0.0005657s
info: GetState[0]
      Function GetState invoking.
info: GetState[0]
      Function GetState succeeded.
info: GetState[0]
      Function completed. Duration: 7.9E-05s
info: ChangeState[0]
      Function ChangeState invoking.
[Light is now on]
info: ChangeState[0]
      Function ChangeState succeeded.
info: ChangeState[0]
      Function completed. Duration: 0.0007424s
fail: Microsoft.SemanticKernel.Planning.Handlebars.HandlebarsPlan[0]
      Plan creation failed. Error: [HallucinatedHelpers] The plan references hallucinated helpers: Helper 'if'
      Microsoft.SemanticKernel.KernelException: [HallucinatedHelpers] The plan references hallucinated helpers: Helper 'if'
       ---> HandlebarsDotNet.HandlebarsRuntimeException: Template references a helper that cannot be resolved. Helper 'if'
         at HandlebarsDotNet.Helpers.MissingHelperDescriptor.Invoke(HelperOptions& options, Context& context, Arguments& arguments)
         at HandlebarsDotNet.Helpers.MissingHelperDescriptor.HandlebarsDotNet.Helpers.IHelperDescriptor<HandlebarsDotNet.HelperOptions>.Invoke(HelperOptions& options, Context& context, Arguments& arguments)
         at HandlebarsDotNet.Helpers.LateBindHelperDescriptor.Invoke(HelperOptions& options, Context& context, Arguments& arguments)
         at HandlebarsDotNet.Helpers.LateBindHelperDescriptor.HandlebarsDotNet.Helpers.IHelperDescriptor<HandlebarsDotNet.HelperOptions>.Invoke(HelperOptions& options, Context& context, Arguments& arguments)
         at lambda_method13(Closure, EncodedTextWriter&, BindingContext)
         at HandlebarsDotNet.HandlebarsEnvironment.<>c__DisplayClass19_0.<Compile>b__0(TextWriter writer, Object context, Object data)
         at HandlebarsDotNet.HandlebarsEnvironment.<>c__DisplayClass20_0.<Compile>b__0(Object context, Object data)
         at Microsoft.SemanticKernel.PromptTemplates.Handlebars.HandlUnhandled exception. ebarsPromptTemplate.RenderAsync(Kernel kernel, KernelArguments arguments, CancellationToken cancellationToken)
         at Microsoft.SemanticKernel.Planning.Handlebars.HandlebarsPlan.InvokeCoreAsync(Kernel kernel, KernelArguments arguments, CancellationToken cancellationToken)
         --- End of inner exception stack trace ---
         at Microsoft.SemanticKernel.Planning.Handlebars.HandlebarsPlan.InvokeCoreAsync(Kernel kernel, KernelArguments arguments, CancellationToken cancellationToken)
         at Microsoft.SemanticKernel.Planning.PlannerInstrumentation.InvokePlanAsync[TPlan,TPlanInput,TPlanResult](Func`5 InvokePlanAsync, TPlan plan, Kernel kernel, TPlanInput input, ILogger logger, CancellationToken cancellationToken)
info: Microsoft.SemanticKernel.Planning.Handlebars.HandlebarsPlan[0]
      Plan creation duration: 0.0336607s.
Microsoft.SemanticKernel.KernelException: [HallucinatedHelpers] The plan references hallucinated helpers: Helper 'if'
 ---> HandlebarsDotNet.HandlebarsRuntimeException: Template references a helper that cannot be resolved. Helper 'if'
   at HandlebarsDotNet.Helpers.MissingHelperDescriptor.Invoke(HelperOptions& options, Context& context, Arguments& arguments)
   at HandlebarsDotNet.Helpers.MissingHelperDescriptor.HandlebarsDotNet.Helpers.IHelperDescriptor<HandlebarsDotNet.HelperOptions>.Invoke(HelperOptions& options, Context& context, Arguments& arguments)
   at HandlebarsDotNet.Helpers.LateBindHelperDescriptor.Invoke(HelperOptions& options, Context& context, Arguments& arguments)
   at HandlebarsDotNet.Helpers.LateBindHelperDescriptor.HandlebarsDotNet.Helpers.IHelperDescriptor<HandlebarsDotNet.HelperOptions>.Invoke(HelperOptions& options, Context& context, Arguments& arguments)
   at lambda_method13(Closure, EncodedTextWriter&, BindingContext)
   at HandlebarsDotNet.HandlebarsEnvironment.<>c__DisplayClass19_0.<Compile>b__0(TextWriter writer, Object context, Object data)
   at HandlebarsDotNet.HandlebarsEnvironment.<>c__DisplayClass20_0.<Compile>b__0(Object context, Object data)
   at Microsoft.SemanticKernel.PromptTemplates.Handlebars.HandlebarsPromptTemplate.RenderAsync(Kernel kernel, KernelArguments arguments, CancellationToken cancellationToken)
   at Microsoft.SemanticKernel.Planning.Handlebars.HandlebarsPlan.InvokeCoreAsync(Kernel kernel, KernelArguments arguments, CancellationToken cancellationToken)
   --- End of inner exception stack trace ---
   at Microsoft.SemanticKernel.Planning.Handlebars.HandlebarsPlan.InvokeCoreAsync(Kernel kernel, KernelArguments arguments, CancellationToken cancellationToken)
   at Microsoft.SemanticKernel.Planning.PlannerInstrumentation.InvokePlanAsync[TPlan,TPlanInput,TPlanResult](Func`5 InvokePlanAsync, TPlan plan, Kernel kernel, TPlanInput input, ILogger logger, CancellationToken cancellationToken)
```

**Comment by jd-aguilar** [comment]
i had the same error but using another NuGet packages:
```
    <PackageReference Include="Microsoft.SemanticKernel" Version="1.3.0" />
    <PackageReference Include="Microsoft.SemanticKernel.Planners.Handlebars" Version="1.3.0-preview" />
```
Plan and Error: 

`Original plan:
{{!-- Step 1: Set a variable to determine if the user is asking for a form --}}
{{set "isAskingForForm" (equals userMessage "i want to fill the form GoSystem user access")}}

{{!-- Step 2: Extract user name and form name from chat history - For simplicity, hardcoding it as the template doesn't include data extraction logic --}}
{{set "userName" "foxes"}}
{{set "formName" "GoSystem user access"}}

{{!-- Step 3: Retrieve the form JSON based on the form name and user name if there is a specific request for a form, or just by the form name otherwise --}}
{{#with (if isAskingForForm (NewFormsPlugin-GetJsonForm formName=formName) (NewFormsPlugin-GetJsonFormByUserName formName=formName userName=userName))}}
  {{!-- Step 4: Output the retrieved form JSON --}}
  {{json this}}
{{/with}}
[2024-02-08T13:47:15.131Z] Function 'EAMChatBotFunction', Invocation id '45a26c8a-a435-4da3-886e-527eb14a202f': An exception was thrown by the invocation.
[2024-02-08T13:47:15.133Z] Result: Function 'EAMChatBotFunction', Invocation id '45a26c8a-a435-4da3-886e-527eb14a202f': An exception was thrown by the invocation.
Exception: Microsoft.SemanticKernel.KernelException: [HallucinatedHelpers] The plan references hallucinated helpers: Helper 'if'
[2024-02-08T13:47:15.134Z]  ---> HandlebarsDotNet.HandlebarsRuntimeException: Template references a helper that cannot be resolved. Helper 'if'`   

this error also happen if the planner uses contains, dict, lowercase, concat and slice

**Comment by teresaqhoang** [comment]
Hi @godefroi, @jd-aguilar, sorry to hear you're running into errors! These exceptions are being caused by faulty syntax.

HB block helpers like `#if` and `#unless` must follow the specific syntax, in which the opening block must start with a `#` and the closing block starts with `/`, i.e., 
```handlebars
{{#if someCondition}} Do Something {{/if}}
```

In the examples you showed above `unless` and `if` are missing the `#` when being invoked. 
I added a minor prompt change in this PR to signal the model to be more careful with syntax: https://github.com/microsoft/semantic-kernel/pull/5137/files

Otherwise, your best bet is to re-run the planner when seeing these faulty plans. Please note that the planner works best with gpt-4. Thanks!

Closing this as resolved for now, please re-open if you guys are still seeing issues.

**Comment by godefroi** [comment]
@kosimas I'm using Mistral-7b-Instruct. @teresaqhoang I'll test this when it becomes available; as I'm using LlamaSharp, it could be some time before an updated release references the next release of SK. Thanks for taking a look at this!

**Comment by teresaqhoang** [comment]
Hi @godefroi, the changes are now available as part of release dotnet-1.5.0 https://github.com/microsoft/semantic-kernel/releases/tag/dotnet-1.5.0 

Will close this as resolved, feel free to re-open if you're still seeing the issue persist consistently.

**Comment by godefroi** [comment]
I can confirm that this now works for my case. The LLM planning functionality seems to be *very* sensitive to the particulars of the description of the plugins, which is fair, and is called out in the documentation. Thanks!
